File: front/hotel-frontend/src/api/reservations.ts
import http from './http';
import type {
    ReservationDto,
    GetReservationsQuery,
    CreateReservationCommand,
    UpdateReservationCommand
} from './types/reservations';

/**
 * Reservations API wrapper.
 * Following backend OpenAPI/Clean Architecture endpoints in Web/Endpoints/Reservations.cs
 */

export const searchReservations = async (params: GetReservationsQuery): Promise<ReservationDto[]> => {
    const response = await http.get<ReservationDto[]>('reservations', { params });
    return response.data;
};

export const getReservationDetails = async (id: number): Promise<ReservationDto> => {
    const response = await http.get<ReservationDto>(`reservations/${id}`);
    return response.data;
};

export const createReservation = async (command: CreateReservationCommand): Promise<ReservationDto> => {
    const response = await http.post<ReservationDto>('reservations', command);
    return response.data;
};

export const updateReservation = async (id: number, command: UpdateReservationCommand): Promise<void> => {
    await http.put(`reservations/${id}`, command);
};

export const confirmReservation = async (id: number): Promise<void> => {
    await http.post(`reservations/${id}/confirm`);
};

export const checkInReservation = async (id: number): Promise<void> => {
    await http.post(`reservations/${id}/checkin`);
};

export const checkOutReservation = async (id: number): Promise<void> => {
    await http.post(`reservations/${id}/checkout`);
};

export const noShowReservation = async (id: number): Promise<void> => {
    await http.post(`reservations/${id}/noshow`);
};

export const cancelReservation = async (id: number): Promise<void> => {
    await http.post(`reservations/${id}/cancel`, {});
};

export const deleteReservation = async (id: number, reason?: string): Promise<void> => {
    await http.delete(`reservations/${id}`, { params: { reason } });
};

File: backend/src/Web/Endpoints/Reservations.cs
using CleanArchitecture.Application.Reservations.Commands.CancelReservation;
using CleanArchitecture.Application.Reservations.Commands.CheckInReservation;
using CleanArchitecture.Application.Reservations.Commands.CheckOutReservation;
using CleanArchitecture.Application.Reservations.Commands.ConfirmReservation;
using CleanArchitecture.Application.Reservations.Commands.CreateReservation;
using CleanArchitecture.Application.Reservations.Commands.DeleteReservation;
using CleanArchitecture.Application.Reservations.Commands.NoShowReservation;
using CleanArchitecture.Application.Reservations.Commands.UpdateReservation;
using CleanArchitecture.Application.Reservations.Queries.GetReservationById;
using CleanArchitecture.Application.Reservations.Queries.GetReservations;
using Microsoft.AspNetCore.Http.HttpResults;

namespace CleanArchitecture.Web.Endpoints;

public class Reservations : EndpointGroupBase
{
    public override string? GroupName => "reservations";

    public override void Map(RouteGroupBuilder group)
    {
        group.RequireAuthorization();

        group.MapGet("", GetReservations);
        group.MapGet("{id}", GetReservationById);
        group.MapPost(CreateReservation);
        group.MapPut("{id}", UpdateReservation);
        group.MapPost("{id}/confirm", ConfirmReservation);
        group.MapPost("{id}/checkin", CheckInReservation);
        group.MapPost("{id}/checkout", CheckOutReservation);
        group.MapPost("{id}/noshow", NoShowReservation);
        group.MapPost("{id}/cancel", CancelReservation);
        group.MapDelete("{id}", DeleteReservation);
    }

    public async Task<Ok<List<ReservationDto>>> GetReservations(ISender sender, [AsParameters] GetReservationsQuery query)
    {
        return TypedResults.Ok(await sender.Send(query));
    }

    public async Task<Results<Ok<ReservationDto>, NotFound>> GetReservationById(ISender sender, int id)
    {
        var item = await sender.Send(new GetReservationByIdQuery(id));
        return TypedResults.Ok(item);
    }

    public async Task<Created<ReservationDto>> CreateReservation(ISender sender, CreateReservationCommand command)
    {
        var result = await sender.Send(command);
        return TypedResults.Created($"/api/reservations/{result.Id}", result);
    }

    public async Task<Results<NoContent, BadRequest, NotFound>> UpdateReservation(ISender sender, int id, UpdateReservationCommand command)
    {
        if (id != command.Id) return TypedResults.BadRequest();

        await sender.Send(command);

        return TypedResults.NoContent();
    }

    public async Task<Results<NoContent, NotFound>> ConfirmReservation(ISender sender, int id)
    {
        await sender.Send(new ConfirmReservationCommand(id));
        return TypedResults.NoContent();
    }

    public async Task<Results<NoContent, NotFound>> CheckInReservation(ISender sender, int id)
    {
        await sender.Send(new CheckInReservationCommand(id));
        return TypedResults.NoContent();
    }

    public async Task<Results<NoContent, NotFound>> CheckOutReservation(ISender sender, int id)
    {
        await sender.Send(new CheckOutReservationCommand(id));
        return TypedResults.NoContent();
    }

    public async Task<Results<NoContent, NotFound>> NoShowReservation(ISender sender, int id)
    {
        await sender.Send(new NoShowReservationCommand(id));
        return TypedResults.NoContent();
    }

    public async Task<Results<NoContent, NotFound>> CancelReservation(ISender sender, int id, CancelReservationRequest request)
    {
        await sender.Send(new CancelReservationCommand(id, request.Reason));
        return TypedResults.NoContent();
    }

    public record CancelReservationRequest(string? Reason);

    public async Task<Results<NoContent, NotFound>> DeleteReservation(ISender sender, int id, string? reason)
    {
        await sender.Send(new DeleteReservationCommand(id, reason));
        return TypedResults.NoContent();
    }
}
